src/uploads/imageResizer.ts

```diff
@@ -6,128 +6,212 @@ import sharp from 'sharp';
+ * @returns the sanitized name and extension
+ */
+const getSanitizedImageData = (sourceImage: string): SanitizedImageData => {
   const extension = sourceImage.split('.').pop();
   const name = sanitize(sourceImage.substring(0, sourceImage.lastIndexOf('.')) || sourceImage);
-
-  return {
-    name,
-    extension,
-    width: size.width,
-    height: size.height,
-  };
-}
-
+  return { name, ext: extension! };
+};
+
+/**
+ * Create a new image name based on the output image name, the dimensions and
+ * the extension.
+ *
+ * Ignore the fact that duplicate names could happen if the there is one
+ * size with `width AND height` and one with only `height OR width`. Because
+ * space is expensive, we will reuse the same image for both sizes.
+ *
+ * @param sanitizedImage - the sanitized image name
+ * @param bufferInfo - the buffer info
+ * @param extension - the extension to use
+ * @param takenOutputNames - the taken output names
+ * @returns the new image name that is not taken
+ */
+const createImageName = (
+  outputImageName: string,
+  { width, height }: sharp.OutputInfo,
+  extension: string,
+) => `${outputImageName}-${width}x${height}.${extension}`;
+
+/**
+ * Create the result object for the image resize operation based on the
+ * provided parameters. If the name is not provided, an empty result object
+ * is returned.
+ *
+ * @param name - the name of the image
+ * @param filename - the filename of the image
+ * @param width - the width of the image
+ * @param height - the height of the image
+ * @param filesize - the filesize of the image
+ * @param mimeType - the mime type of the image
+ * @param sizesToSave - the sizes to save
+ * @returns the result object
+ */
+const createResult = (
+  name: string,
+  filename: FileSize['filename'] = null,
+  width: FileSize['width'] = null,
+  height: FileSize['height'] = null,
+  filesize: FileSize['filesize'] = null,
+  mimeType: FileSize['mimeType'] = null,
+  sizesToSave: FileToSave[] = [],
+): ResizeResult => ({
+  sizesToSave,
+  sizeData: {
+    [name]: {
+      width,
+      height,
+      filename,
+      filesize,
+      mimeType,
+    },
+  },
+});
+
+/**
+ * Check if the image needs to be resized according to the requested dimensions
+ * and the original image size. If the resize options are provided, the image
+ * will be resized regardless of the requested dimensions, given that the
+ * width or height to be resized is provided.
+ *
+ * @param requestedDimensions - the requested dimensions
+ * @param original - the original image size
+ * @param resizeOptions - the resize options
+ * @returns true if the image needs to be resized, false otherwise
+ */
+const needsResize = (
+  { width, height, withoutEnlargement, withoutReduction }: ImageSize,
+  original: ProbedImageSize,
+): boolean => {
+  // allow enlargement or prevent reduction (our default is to prevent
+  // enlargement and allow reduction)
+  if (withoutEnlargement === false || withoutReduction) {
```

Comment:
> If `withoutReduction` is set to false, won't we need a check which verifies that the requested image dimensions are higher than the original image dimensions?

That would be the default behavior which you can find in the rest of the function. 

https://github.com/payloadcms/payload/pull/2844/files/d5da0524f42a2e7661e5fb1a7bea78c5a7fc4a63#diff-a7622882270f6b696698fb27dbcb8644babad703c75164cc230d711e7b176c3cR119-R125

I didn't want to change the default behavior that is currently available. Therefore we only `force` a resize if a user decides to use `withoutEnlargement` or `withoutReduction`. The properties are described as the following, but we never come to the point that `withoutEnlargement` == `true` because we check if the image has to be resized beforehand.

Param | Type | Default | Description
-- | -- | -- | --
[options.withoutEnlargement] | Boolean | false | Do not scale up if the width or height are already less than the target dimensions.
[options.withoutReduction] | Boolean | false | Do not scale down if the width or height are already greater than the target dimensions.

> Also, do you think a check where it checks if width & height are equal to the original width and height would be helpful here? That way, we wouldn't unnecessarily resize those.

I checked this behavior too and also implemented logic that checked if the image was actually resized or had a format change via `formatOptions`. But that felt too complicated and bloated the whole functionality. Further, I checked if the `srcset` will recognize if one of the image URLs is null or not available and will use the original. Sadly, it did not, so I think we should provide the image anyway to reduce errors on users end. Since the user has to willingly deactivate `withoutEnlargement`, it is also much less error prone.

 What do you think about that?

Reply:
> I checked this behavior too and also implemented logic that checked if the image was actually resized or had a format change via formatOptions. But that felt too complicated and bloated the whole functionality. 

Wouldn't a simple

```ts
if (height === original.height && width === original.width) {
    return false;
  }
```

in our `needsResize` method do the job?


> That would be the default behavior which you can find in the rest of the function.
> 
> https://github.com/payloadcms/payload/pull/2844/files/d5da0524f42a2e7661e5fb1a7bea78c5a7fc4a63#diff-a7622882270f6b696698fb27dbcb8644babad703c75164cc230d711e7b176c3cR119-R125
> 
> I didn't want to change the default behavior that is currently available. Therefore we only force a resize if a user decides to use withoutEnlargement or withoutReduction. The properties are described as the following, but we never come to the point that withoutEnlargement == true because we check if the image has to be resized beforehand.

I've actually completely disregarded that sharp probably by default doesn't reduce the image if it is already smaller than the specified sizes, so whoops, nevermind that.

Could you explain why we need the `|| withoutReduction` in the `needsResize` method in the first place? Not quite sure if I understand that.

Reply:
> Wouldn't a simple {code} in our needsResize method do the job?

Not quite, because you also need to check the requested format in that case.
If the format differs, you want sharp to do the job. If it is the same, you could check it that way.


> Could you explain why we need the `|| withoutReduction` in the `needsResize` method in the first place? Not quite sure if I understand that.

Just to support the property if someone is willingly setting it. And there I entirely ignore if it is required to do a resize, since it is non default.